### ISSUE TITLE
support AUTOSAR C++14 A7-5-1 bugprone-return-const-ref-from-parameter

### ISSUE THREAD

**Issue by HerrCai0907** [llvm/llvm-project#85253]
> AUTOSAR C++14 A7-5-1 
> A function shall not return a reference or a pointer to a parameter that is passed by reference to const.

[example](https://godbolt.org/z/9zMEco9E3)

```c++
struct S {
    S();
    S(S const&);
    ~S();
};

S const& f(S const &a) { return a; }


S f() {
    S const & a = f(S{}); // object referenced by 'a' has destructed after this statement
    return a; // bugprone
}
```

**Comment by sopyb** [comment]
I began working on this check request. I am using [AUTOSAR CPP14 Guidelines](https://www.autosar.org/fileadmin/standards/R22-11/AP/AUTOSAR_RS_CPP14Guidelines.pdf) (pages 147-149) as a reference for the new check's functionality. I expect to have a pull request ready by Wednesday, possibly sooner if I don't face any unforeseen issues. 

Any other resources I could reference would be greatly appreciated.

**Comment by PiotrZSL** [comment]
I think this check should be regardless this is const reference, reference or rvalue reference.

```
S&& f(S &&a) { return std::move(a); }

S const & a = f(S{}); // object referenced by 'a' has destructed after this statement
```

check should detect functions that return reference to argument.

**Comment by carlosgalvezp** [comment]
Please note: we are not allowed to implement AUTOSAR rules:

https://discourse.llvm.org/t/clang-tidy-rfc-add-autosar-c-14-clang-tidy-module/59223

Implementing this functionality without using AUTOSAR as a reference would be fine, though. Is there a similar e.g. C++ Core Guideline that could be used instead?

**Comment by HerrCai0907** [comment]
> Please note: we are not allowed to implement AUTOSAR rules:
> 
> 
> 
> https://discourse.llvm.org/t/clang-tidy-rfc-add-autosar-c-14-clang-tidy-module/59223
> 
> https://reviews.llvm.org/D112730
> 
> 
> 
> Implementing this functionality without using AUTOSAR as a reference would be fine, though. Is there a similar e.g. C++ Core Guideline that could be used instead?

 Could we avoid legacy issues if we don't mention any autosar related things?

**Comment by carlosgalvezp** [comment]
> Could we avoid legacy issues if we don't mention any autosar related things?

Sure, if we can avoid any mention to AUTOSAR and not copy anything from that document (rule text, rationale, code examples) then I think it should be good.
